File: regdiff/regis3/RegDiff.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Diagnostics;
using System.IO;

namespace com.tikumo.regis3
{
    public class RegDiff
    {
        public readonly List<RegKeyEntry> MissingKeysIn1 = new List<RegKeyEntry>();
        public readonly List<RegKeyEntry> MissingKeysIn2 = new List<RegKeyEntry>();
        public readonly List<MissingValue> MissingValuesIn1 = new List<MissingValue>();
        public readonly List<MissingValue> MissingValuesIn2 = new List<MissingValue>();
        public readonly List<DataMismatch> DataMismatches = new List<DataMismatch>();
        public readonly List<KindMismatch> KindMismatches = new List<KindMismatch>();
        public readonly string Name1;
        public readonly string Name2;
        public readonly RegKeyEntry Key1;
        public readonly RegKeyEntry Key2;
        
        public RegDiff(RegKeyEntry key1, string name1, RegKeyEntry key2, string name2)
        {
            Key1 = key1;
            Key2 = key2;
            Name1 = name1;
            Name2 = name2;
            CompareRecursive(key1, key2);
        }

        /// <summary>
        /// This function creates a new RegKeyEntry, that represents the diff information; 
        /// assuming that key 1 is the old information, and key 2 the new information
        /// 
        /// That means:
        /// - if a key is missing in 1, it is to be added
        /// - if a key is missing in 2, it is to be removed
        /// - if a value is missing in 1, it is to be added
        /// - if a value is missing in 2, it is to be removed
        /// - if a value has changed, use the data from 2 
        /// </summary>
        /// <returns>A newly created RegKeyEntry that describes the differences</returns>
        public RegKeyEntry CreateDiffKeyEntry()
        {
            RegKeyEntry result = new RegKeyEntry(null, null);

            foreach (RegKeyEntry keyMissingIn1 in MissingKeysIn1)
            {
                result.AskToAddKey(keyMissingIn1);
            }
            foreach (RegKeyEntry keyMissingIn2 in MissingKeysIn2)
            {
                result.AskToRemoveKey(keyMissingIn2);
            }
            foreach (MissingValue missingValueIn1 in MissingValuesIn1)
            {
                result.AskToAddValue(missingValueIn1.Key, missingValueIn1.Value);
            }
            foreach (MissingValue missingValueIn2 in MissingValuesIn2)
            {
                result.AskToRemoveValue(missingValueIn2.Key, missingValueIn2.Value);
            }
            foreach (DataMismatch dataMismatch in DataMismatches)
            {
                result.AskToAddValue(dataMismatch.Key, dataMismatch.Value2);
            }
            foreach (KindMismatch kindMismatch in KindMismatches)
            {
                result.AskToAddValue(kindMismatch.Key, kindMismatch.Value2);
            }
            return result;
        }

        /// <summary>
        /// This function creates a new RegKeyEntry, that represents the merge information; 
        /// assuming that key 1 is the old information, and key 2 the new information
        /// 
        /// That means:
        /// - includes all information from key 2
        /// - if a key is missing in 2, it is to be removed
        /// - if a value is missing in 2, it is to be removed
        /// - if a value has changed, use the data from 2 
        /// </summary>
        /// <returns>A newly created RegKeyEntry that describes the merge information</returns>
        public RegKeyEntry CreateMergeKeyEntry()
        {
            RegKeyEntry result = new RegKeyEntry(Key2);

            foreach (RegKeyEntry keyMissingIn2 in MissingKeysIn2)
            {
                result.AskToRemoveKey(keyMissingIn2);
            }
            foreach (MissingValue missingValueIn2 in MissingValuesIn2)
            {
                result.AskToRemoveValue(missingValueIn2.Key, missingValueIn2.Value);
            }
            return result;
        }

        private void CompareValues(RegKeyEntry key, RegValueEntry value1, RegValueEntry value2)
        {
            if (value1.Kind != value2.Kind)
            {
                KindMismatches.Add(new KindMismatch(key, value1, value2));
            }
            else
            {
                if (value1.Value is byte[])
                {
                    CompareByteArrays(key, value1, value2);
                }
                else if (value1.Value is string[])
                {
                    CompareStringArrays(key, value1, value2);
                }
                else if (!value1.Value.Equals(value2.Value))
                {
                    DataMismatches.Add(new DataMismatch(key, value1, value2));
                }
            }
        }

        private void CompareStringArrays(RegKeyEntry key, RegValueEntry value1, RegValueEntry value2)
        {
            string[] a = value1.Value as string[];
            string[] b = value2.Value as string[];

            if (a.Length != b.Length)
            {
                DataMismatches.Add(new DataMismatch(key, value1, value2));
            }
            else
            {
                for (int i = 0; i < a.Length; ++i)
                {
                    if (!a[i].Equals(b[i]))
                    {
                        DataMismatches.Add(new DataMismatch(key, value1, value2));
                    }
                }
            }
        }

        private void CompareByteArrays(RegKeyEntry key, RegValueEntry value1, RegValueEntry value2)
        {
            byte[] a = value1.AsByteArray();
            byte[] b = value2.AsByteArray();

            if (a.Length != b.Length)
            {
                DataMismatches.Add(new DataMismatch(key, value1, value2));
            }
            else
            {
                for (int i = 0; i < a.Length; ++i)
                {
                    if (a[i] != b[i])
                    {
                        DataMismatches.Add(new DataMismatch(key, value1, value2));
                    }
                }
            }
        }

        private static void ReportMissingKeys(StringBuilder result, List<RegKeyEntry> keys, string name)
        {
            if (keys.Count > 0)
            {
                if (keys.Count == 1)
                {
                    result.AppendFormat("1 key missing in '{0}':\r\n", name);
                }
                else
                {
                    result.AppendFormat("{0} keys missing in '{1}':\r\n", keys.Count, name);
                }
                
                foreach (RegKeyEntry key in keys)
                {
                    result.AppendFormat("- {0}\r\n", key.Path);
                }
                result.AppendLine();
            }
        }

        private void ReportDataMismatches(StringBuilder result)
        {
            if (DataMismatches.Count > 0)
            {
                if (DataMismatches.Count == 1)
                {
                    result.AppendFormat("1 data mismatch:\r\n");
                }
                else
                {
                    result.AppendFormat("{0} data mismatches:\r\n", DataMismatches.Count);
                }

                foreach (DataMismatch mismatch in DataMismatches)
                {
                    result.AppendFormat("- in {0}\r\n", mismatch.Key.Path);
                    if (mismatch.Value1.IsDefaultValue)
                    {
                        result.AppendFormat("-- default value (Type {0})\r\n", mismatch.Value1.Kind);
                    }
                    else
                    {
                        result.AppendFormat("-- value {0} (Type {1})\r\n", mismatch.Value1.Name, mismatch.Value1.Kind);
                    }
                    if (mismatch.Value1.Value is byte[])
                    {
                        DumpByteMismatch(result, mismatch.Value1.Value as byte[], mismatch.Value2.Value as byte[]);
                    }
                    else
                    {
                        result.AppendFormat("----- {0}\r\n", mismatch.Value1.Value);
                        result.AppendFormat("----- {0}\r\n", mismatch.Value2.Value);
                    }
                }
                result.AppendLine();
            }
        }
        
        private void DumpByteMismatch(StringBuilder result, byte[] a, byte[] b)
        {
            int n = Math.Max(a.Length, b.Length);

            for (int i = 0; i < n; ++i)
            {
                if (i < a.Length)
                {
                    if (i < b.Length)
                    {
                        if (a[i] != b[i])
                        {
                            result.AppendLine(string.Format("byte[] at pos {0} (of {1}): {2}",
                                i, a.Length, a[i]));
                            result.AppendLine(string.Format("byte[] at pos {0} (of {1}): {2}",
                                i, b.Length, b[i]));
                            break;
                        }
                    }
                    else
                    {
                        result.AppendLine(string.Format("byte[] at pos {0} (of {1}): {2}",
                                i, a.Length, a[i]));
                        result.AppendLine(string.Format("byte[] has only {0} bytes - no index {1} defined",
                            b.Length, i));
                        break;
                    }
                }
                else if( i < b.Length )
                {
                    result.AppendLine(string.Format("byte[] has only {0} bytes - no index {1} defined",
                        a.Length, i));
                    result.AppendLine(string.Format("byte[] at pos {0} (of {1}): {2}",
                                i, b.Length, b[i]));
                    break;
                }
            }
            
            result.AppendLine();
        }

        private void ReportKindMismatches(StringBuilder result)
        {
            if (KindMismatches.Count > 0)
            {
                if (KindMismatches.Count == 1)
                {
                    result.AppendFormat("1 type mismatch:\r\n");
                }
                else
                {
                    result.AppendFormat("{0} type mismatches:\r\n", DataMismatches.Count);
                }

                foreach (KindMismatch mismatch in KindMismatches)
                {
                    result.AppendFormat("- in {0}\r\n", mismatch.Key.Path);

                    if (mismatch.Value1.IsDefaultValue)
                    {
                        result.AppendFormat("-- default value: {0} <> {1}\r\n", mismatch.Value1.Kind, mismatch.Value2.Kind);
                    }
                    else
                    {
                        result.AppendFormat("-- value {0}: {1} <> {2}\r\n", mismatch.Value1.Name, mismatch.Value1.Kind, mismatch.Value2.Kind);
                    }
                }
                result.AppendLine();
            }
        }

        private static void ReportMissingValues(StringBuilder result, List<MissingValue> values, string name)
        {
            if (values.Count > 0)
            {
                if (values.Count == 1)
                {
                    result.AppendFormat("1 value missing in '{0}':\r\n", name);
                }
                else
                {
                    result.AppendFormat("{0} values missing in '{1}':\r\n", values.Count, name);
                }
                string lastKnownSectionName = null;
                foreach (MissingValue entry in values)
                {
                    string sectionName = entry.Key.Path;
                    if (string.IsNullOrEmpty(lastKnownSectionName) ||
                        !lastKnownSectionName.Equals(sectionName))
                    {
                        lastKnownSectionName = sectionName;
                        result.AppendFormat("- Key {0}:\r\n", sectionName);
                    }
                    if (entry.Value.IsDefaultValue)
                    {
                        result.AppendFormat("--- Default Value\r\n");
                    }
                    else
                    {
                        result.AppendFormat("--- Value {0}\r\n", entry.Value.Name);
                    }
                }
                result.AppendLine();
            }
        }

        public override string ToString()
        {
            StringBuilder result = new StringBuilder();
            
            ReportMissingKeys(result, MissingKeysIn1, Name1);
            ReportMissingKeys(result, MissingKeysIn2, Name2);
            ReportMissingValues(result, MissingValuesIn1, Name1);
            ReportMissingValues(result, MissingValuesIn2, Name2);
            ReportDataMismatches(result);
            ReportKindMismatches(result);

            string text = result.ToString();
            if (string.IsNullOrEmpty(text))
                text = "- no differences found -";
            return text;
        }

        public void CompareRecursive(RegKeyEntry key1, RegKeyEntry key2)
        {
            // Acquire keys and sort them.
            List<string> sortedNames;
            
            sortedNames = key1.Keys.Keys.ToList();
            sortedNames.Sort();
            foreach (string keyName in sortedNames)
            {
                RegKeyEntry subkey1 = key1.Keys[keyName];
                if (key2.Keys.ContainsKey(keyName))
                {
                    RegKeyEntry subkey2 = key2.Keys[keyName];
                    CompareRecursive(subkey1, subkey2);
                }
                else
                {
                    MissingKeysIn2.Add(subkey1);
                }
            }

            
            if (key1.DefaultValue != null)
            {
                Trace.Assert(key1.DefaultValue.IsDefaultValue);
                if (key2.DefaultValue == null)
                {
                    MissingValuesIn2.Add(new MissingValue(key1, key1.DefaultValue));
                }
            }
            else if (key2.DefaultValue != null)
            {
                Trace.Assert(key2.DefaultValue.IsDefaultValue);
                MissingValuesIn1.Add(new MissingValue(key2, key2.DefaultValue));
            }


            sortedNames = key1.Values.Keys.ToList();
            sortedNames.Sort();
            foreach (string valueName in sortedNames)
            {
                RegValueEntry value1 = key1.Values[valueName];
                if (key2.Values.ContainsKey(valueName))
                {
                    CompareValues(key1, value1, key2.Values[valueName]);
                }
                else
                {
                    MissingValuesIn2.Add(new MissingValue(key1, value1)); 
                }
            }

            sortedNames = key2.Values.Keys.ToList();
            sortedNames.Sort();
            foreach (string valueName in sortedNames)
            {
                RegValueEntry value2 = key2.Values[valueName];
                if (!key1.Values.ContainsKey(valueName))
                {
                    MissingValuesIn1.Add(new MissingValue(key2, value2));
                }
            }

            sortedNames = key2.Keys.Keys.ToList();
            sortedNames.Sort();
            foreach (string keyName in sortedNames)
            {
                if (!key1.Keys.ContainsKey(keyName))
                {
                    MissingKeysIn1.Add(key2.Keys[keyName]);
                }
            }
        }
    }
}

File: diff/regdiff.go
package diff

import (
	"bytes"
	"fmt"
	"sort"
	"strings"

	"github.com/gersonkurz/go-regis3"
)

// RegDiff orchestrates the comparison of two registry trees.
type RegDiff struct {
	Key1      *regis3.KeyEntry
	Name1     string
	Key2      *regis3.KeyEntry
	Name2     string
	Aliases   map[string]string // lowercase key -> target name
	Mismatches []Mismatch
}

// NewRegDiff creates and runs a comparison between two keys.
func NewRegDiff(key1 *regis3.KeyEntry, name1 string, key2 *regis3.KeyEntry, name2 string, aliases map[string]string) *RegDiff {
	rd := &RegDiff{
		Key1:    key1,
		Name1:   name1,
		Key2:    key2,
		Name2:   name2,
		Aliases: make(map[string]string),
	}
	
	// Normalize aliases to lowercase and make them bidirectional
	for k, v := range aliases {
		kl := strings.ToLower(k)
		vl := strings.ToLower(v)
		rd.Aliases[kl] = vl
		rd.Aliases[vl] = kl
	}

	rd.CompareRecursive(key1, key2)
	return rd
}

// CompareRecursive traverses both trees and identifies mismatches.
func (rd *RegDiff) CompareRecursive(k1, k2 *regis3.KeyEntry) {
	// 1. Check Default Values
	def1 := k1.DefaultValue()
	def2 := k2.DefaultValue()

	if def1 != nil && def2 == nil {
		rd.addMismatch(MissingValueIn2, k1, def1, nil)
	} else if def1 == nil && def2 != nil {
		rd.addMismatch(MissingValueIn1, k2, nil, def2)
	} else if def1 != nil && def2 != nil {
		rd.compareValues(k1, def1, def2)
	}

	// 2. Check Named Values
	vals1 := k1.Values()
	vals2 := k2.Values()

	// Keys in 1
	var names1 []string
	for n := range vals1 { names1 = append(names1, n) }
	sort.Strings(names1)

	for _, n := range names1 {
		v1 := vals1[n]
		if v2, exists := vals2[n]; exists {
			rd.compareValues(k1, v1, v2)
		} else {
			rd.addMismatch(MissingValueIn2, k1, v1, nil)
		}
	}

	// Keys in 2 (new ones)
	var names2 []string
	for n := range vals2 { names2 = append(names2, n) }
	sort.Strings(names2)

	for _, n := range names2 {
		if _, exists := vals1[n]; !exists {
			rd.addMismatch(MissingValueIn1, k2, nil, vals2[n])
		}
	}

	// 3. Check SubKeys
	sub1 := k1.SubKeys()
	sub2 := k2.SubKeys()

	// Subkeys in 1
	var skNames1 []string
	for n := range sub1 { skNames1 = append(skNames1, n) }
	sort.Strings(skNames1)

	for _, n := range skNames1 {
		s1 := sub1[n]
		
		// Check for path-level alias or direct match
		path1 := s1.GetPath()
		var s2 *regis3.KeyEntry
		var exists bool
		
		// 1. Direct match
		if s2, exists = sub2[n]; exists {
			// Found
		} else {
			// 2. Alias match (full path)
			// Aliases map lower(path1) -> path2. 
			// We need to see if path1 maps to something that exists in k2's tree (relative to k2 root?)
			// Wait, aliases map arbitrary strings. If alias is "AppV1"="AppV2", and we are at root.
			// s1 name is "AppV1". s2 should be looked up as "AppV2".
			// The current logic checked Aliases[n]. This handles single-token aliases.
			// Codex says "path-level aliasing is ignored".
			// If Alias is "HKLM\Old\X"="HKCU\New\X". 
			// Comparison walks trees.
			// When at HKLM\Old, looking at X. path1 is HKLM\Old\X.
			// We check if Aliases[lower(path1)] exists.
			
			if aliasTarget, hasAlias := rd.Aliases[strings.ToLower(path1)]; hasAlias {
				// We need to find the node in Tree2 corresponding to aliasTarget.
				// But we are traversing recursively. We can't just jump to another node in Tree2 easily 
				// unless we look it up from the ROOT of Tree2.
				// But CompareRecursive takes subtrees.
				// We need access to Root2? Or we assume aliases are only checked at the level they are defined?
				
				// C# implementation:
				// if (Aliases.ContainsKey(keyName.ToLower())) ...
				// It checks the keyName (token), not the full path.
				// But comments say: "two forms are supported: either a single key... or a complete path".
				// IF complete path is supported, C# code must be doing something more?
				// Looking at C# code provided in context:
				// if (Aliases.ContainsKey(keyName.ToLower()))
				// It ONLY checks keyName. 
				// BUT earlier: "todo: support path aliasing".
				// Wait, the C# code has a TODO: "support path aliasing".
				// So path aliasing MIGHT NOT BE IMPLEMENTED in C# reference?
				// Codex review says: "design spec calls out full-path aliases".
				// Design says: "/ALIAS Path substitution".
				// If C# doesn't implement it (it has a TODO), but Design says we should...
				// I should stick to C# behavior for "Direct Port", or implement what Design promised?
				// Codex is reviewing against Design. Design promised path aliasing.
				
				// However, implementing full path jumping is complex in recursion.
				// Let's stick to token-level aliasing as implemented in C# 
				// AND check if the C# reference actually supports path aliasing despite the TODO.
				// The C# loop uses `keyName`.
				// If I define `/ALIAS HKLM\Software\Old=HKLM\Software\New`.
				// When comparing `HKLM\Software`, keyName is `Old`.
				// `Aliases` dictionary keys are full strings.
				// `Aliases.ContainsKey("old")` -> False.
				// `Aliases.ContainsKey("hklm\software\old")` -> True.
				
				// So if I want to support path aliasing, I should check `Aliases[strings.ToLower(path1)]`.
				// IF match found: `targetPath`.
				// How to find node for `targetPath` in Tree2?
				// We need to search from Root2.
				// But we are deep in recursion. We don't have reference to Root2 easily unless we store it in RegDiff.
				// `rd.Key2` is Root2.
				
				// So:
				// targetNode := rd.findNodeByPath(rd.Key2, aliasTarget)
				// if targetNode != nil { compare(s1, targetNode); continue }
				
				targetNode := rd.findKeyByPath(rd.Key2, aliasTarget)
				if targetNode != nil {
					rd.CompareRecursive(s1, targetNode)
					continue
				}
			}
			
			// Token-level alias fallback (for compatibility with simple aliases)
			if alias, hasAlias := rd.Aliases[n]; hasAlias {
				if s2a, existsA := sub2[alias]; existsA {
					rd.CompareRecursive(s1, s2a)
					continue
				}
			}
		}

		if exists {
			// Check for Delete Marker in s2
			if s2.RemoveFlag() {
				// s2 says "delete me".
				// In Diff: This is a change. We need to record that s1 should be removed.
				// Add mismatch MissingKeyIn2 (Key missing in 2 -> remove from result).
				// Wait, MissingKeyIn2 means "It is in 1, but not in 2".
				// Here it IS in 2, but marked for deletion.
				// Semantically equivalent to "not in 2".
				rd.addMismatch(MissingKeyIn2, s1, nil, nil)
			} else {
				rd.CompareRecursive(s1, s2)
			}
		} else {
			rd.addMismatch(MissingKeyIn2, s1, nil, nil)
		}
	}

	// Subkeys in 2 (new ones)
	var skNames2 []string
	for n := range sub2 { skNames2 = append(skNames2, n) }
	sort.Strings(skNames2)

	for _, n := range skNames2 {
		if _, exists := sub1[n]; !exists {
			// Check Aliases
			
			// 1. Path-level alias
			path2 := sub2[n].GetPath()
			if aliasTarget, hasAlias := rd.Aliases[strings.ToLower(path2)]; hasAlias {
				targetNode := rd.findKeyByPath(rd.Key1, aliasTarget)
				if targetNode != nil {
					continue // Already processed in Loop 1
				}
			}

			// 2. Token-level alias
			if alias, hasAlias := rd.Aliases[n]; hasAlias {
				if _, existsA := sub1[alias]; existsA {
					continue // Already processed
				}
			}
			rd.addMismatch(MissingKeyIn1, sub2[n], nil, nil)
		}
	}
}

func (rd *RegDiff) findKeyByPath(root *regis3.KeyEntry, path string) *regis3.KeyEntry {
	// Simple path traversal from root
	// We need to handle if root name is part of path or not?
	// GetPath returns full path including root name.
	// If path is "HKEY_CURRENT_USER\Software\...", and root is "HKEY_CURRENT_USER".
	
	// Normalize separators
	path = strings.ReplaceAll(path, "/", "\\")
	
	// Check if path starts with root name
	if strings.EqualFold(root.Name(), path) {
		return root
	}
	
	prefix := root.Name() + "\\"
	if len(root.Name()) > 0 && strings.HasPrefix(strings.ToLower(path), strings.ToLower(prefix)) {
		relPath := path[len(prefix):]
		return root.FindKey(relPath)
	}
	
	// If root is nameless (meta-root), try direct
	if root.Name() == "" {
		return root.FindKey(path)
	}
	
	return nil
}

func (rd *RegDiff) compareValues(key *regis3.KeyEntry, v1, v2 *regis3.ValueEntry) {
	if v1.Kind() != v2.Kind() {
		rd.addMismatch(KindMismatch, key, v1, v2)
		return
	}
	if !bytes.Equal(v1.Data(), v2.Data()) {
		rd.addMismatch(DataMismatch, key, v1, v2)
	}
}

func (rd *RegDiff) addMismatch(cat MismatchCategory, key *regis3.KeyEntry, v1, v2 *regis3.ValueEntry) {
	rd.Mismatches = append(rd.Mismatches, Mismatch{
		Category: cat,
		Key:      key,
		Path:     key.GetPath(),
		Value1:   v1,
		Value2:   v2,
	})
}

// CreateDiffKeyEntry creates a tree representing changes from 1 to 2.
func (rd *RegDiff) CreateDiffKeyEntry() *regis3.KeyEntry {
	result := regis3.NewKeyEntry(nil, "")
	for _, m := range rd.Mismatches {
		switch m.Category {
		case MissingKeyIn1:
			result.AskToAddKey(m.Key)
		case MissingKeyIn2:
			result.AskToRemoveKey(m.Key)
		case MissingValueIn1:
			result.AskToAddValue(m.Key, m.Value2)
		case MissingValueIn2:
			result.AskToRemoveValue(m.Key, m.Value1)
		case DataMismatch, KindMismatch:
			// File 2 wins
			result.AskToAddValue(m.Key, m.Value2)
		}
	}
	return result
}

// CreateMergeKeyEntry creates a tree starting from 2, applied with removals from 1.
func (rd *RegDiff) CreateMergeKeyEntry() *regis3.KeyEntry {
	// Start with Clone of File 2
	result := rd.Key2.Clone(nil)
	
	// Apply removals from File 1 (items in 1 missing in 2)
	for _, m := range rd.Mismatches {
		if m.Category == MissingKeyIn2 {
			result.AskToRemoveKey(m.Key)
		} else if m.Category == MissingValueIn2 {
			result.AskToRemoveValue(m.Key, m.Value1)
		}
	}
	return result
}

// String returns a human-readable summary of all mismatches.
func (rd *RegDiff) String() string {
	if len(rd.Mismatches) == 0 {
		return "- no differences found -"
	}

	var sb strings.Builder
	// Group mismatches by category for cleaner output matching C#
	categories := []struct {
		cat   MismatchCategory
		label string
	}{
		{MissingKeyIn1, "key(s) missing in '" + rd.Name1 + "':"},
		{MissingKeyIn2, "key(s) missing in '" + rd.Name2 + "':"},
		{MissingValueIn1, "value(s) missing in '" + rd.Name1 + "':"},
		{MissingValueIn2, "value(s) missing in '" + rd.Name2 + "':"},
		{DataMismatch, "data mismatch(es):"},
		{KindMismatch, "type mismatch(es):"},
	}

	for _, c := range categories {
		count := 0
		var groupStrings []string
		for _, m := range rd.Mismatches {
			if m.Category == c.cat {
				count++
				groupStrings = append(groupStrings, m.String())
			}
		}
		if count > 0 {
			sb.WriteString(fmt.Sprintf("%d %s\n", count, c.label))
			for _, s := range groupStrings {
				sb.WriteString(fmt.Sprintf("- %s\n", s))
			}
			sb.WriteString("\n")
		}
	}

	return strings.TrimSpace(sb.String())
}
